### ISSUE TITLE
traffic.sidecar.istio.io/excludeOutboundPorts is ignored

### ISSUE THREAD

**Issue by livius-ungureanu** [istio/istio#38415]
### Bug Description

Problem:
========
It looks like the  `traffic.sidecar.istio.io/excludeOutboundPorts`   annotation is ignored by istio.

I have a  microservice running in a kubernetes (minikube) pod that communicates with PostgreSQL on port 5432.
Istio is installed in the minikube instance.

I need that istio-proxy(envoy) does not intercept the PostgreSQL traffic (i.e.  my microservice  talks directly to PostgreSQL bypassing the sidecar ) but even I have added the the annotation above,   `istio-proxy`  still proxies the PostgreSQL  communication:

```
istio-proxy@cmm-8f8c49f78-7pkvj:/$ netstat -an | grep 5432
tcp        0      0 172.17.0.10:58028       10.10.0.43:5432         ESTABLISHED
tcp        0      0 172.17.0.10:58026       10.10.0.43:5432         ESTABLISHED
```

Relevant config & info:
================

1)Here is how I added the `traffic.sidecar.istio.io/excludeOutboundPorts`   annotation:
----------------------------------------
`kubectl get deployments.apps cmm -o yaml`
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "8"
  creationTimestamp: "2022-04-15T12:32:33Z"
  generation: 8
  name: cmm
  namespace: default
  resourceVersion: "68992"
  uid: 2a39c1e5-7ab1-4b84-b62f-8575ac5802b0
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: cmm
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        kubectl.kubernetes.io/restartedAt: "2022-04-16T23:42:38+03:00"
        traffic.sidecar.istio.io/excludeOutboundPorts: "5432"
      creationTimestamp: null
      labels:
        app: cmm
    spec:
      containers:
      - image: legacymigrate:for_k8
        imagePullPolicy: IfNotPresent
        name: cmm
        ports:
        - containerPort: 8713
          protocol: TCP
        - containerPort: 5432
          name: postgre
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2022-04-15T12:47:16Z"
    lastUpdateTime: "2022-04-16T23:12:43Z"
    message: ReplicaSet "cmm-8f8c49f78" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: "2022-04-16T23:12:48Z"
    lastUpdateTime: "2022-04-16T23:12:48Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 8
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```

2)By looking at my pod definition it looks like the pod has received the iptables details to bypass 5432 port:
---------------------------------------------------------------------

`kubectl  get pod cmm-8f8c49f78-7pkvj -o yaml`
```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubectl.kubernetes.io/default-container: cmm
    kubectl.kubernetes.io/default-logs-container: cmm
    kubectl.kubernetes.io/restartedAt: "2022-04-16T23:42:38+03:00"
    prometheus.io/path: /stats/prometheus
    prometheus.io/port: "15020"
    prometheus.io/scrape: "true"
    sidecar.istio.io/status: '{"initContainers":["istio-init"],"containers":["istio-proxy"],"volumes":["istio-envoy","istio-data","istio-podinfo","istio-token","istiod-ca-cert"],"imagePullSecrets":null,"revision":"default"}'
    traffic.sidecar.istio.io/excludeOutboundPorts: "5432"
  creationTimestamp: "2022-04-16T23:12:40Z"
  generateName: cmm-8f8c49f78-
  labels:
    app: cmm
    pod-template-hash: 8f8c49f78
    security.istio.io/tlsMode: istio
    service.istio.io/canonical-name: cmm
    service.istio.io/canonical-revision: latest
  name: cmm-8f8c49f78-7pkvj
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: cmm-8f8c49f78
    uid: bbbb61d7-c0ab-4537-8635-673ae207f5bd
  resourceVersion: "68988"
  uid: c2401eac-3e1a-4506-b811-ae024cf7e6dd
spec:
  containers:
  - image: legacymigrate:for_k8
    imagePullPolicy: IfNotPresent
    name: cmm
    ports:
    - containerPort: 8713
      protocol: TCP
    - containerPort: 5432
      name: postgre
      protocol: TCP
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-cs7l6
      readOnly: true
  - args:
    - proxy
    - sidecar
    - --domain
    - $(POD_NAMESPACE).svc.cluster.local
    - --proxyLogLevel=warning
    - --proxyComponentLogLevel=misc:error
    - --log_output_level=default:info
    - --concurrency
    - "2"
    env:
    - name: JWT_POLICY
      value: third-party-jwt
    - name: PILOT_CERT_PROVIDER
      value: istiod
    - name: CA_ADDR
      value: istiod.istio-system.svc:15012
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: INSTANCE_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
    - name: SERVICE_ACCOUNT
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.serviceAccountName
    - name: HOST_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.hostIP
    - name: PROXY_CONFIG
      value: |
        {}
    - name: ISTIO_META_POD_PORTS
      value: |-
        [
            {"containerPort":8713,"protocol":"TCP"}
            ,{"name":"postgre","containerPort":5432,"protocol":"TCP"}
        ]
    - name: ISTIO_META_APP_CONTAINERS
      value: cmm
    - name: ISTIO_META_CLUSTER_ID
      value: Kubernetes
    - name: ISTIO_META_INTERCEPTION_MODE
      value: REDIRECT
    - name: ISTIO_META_WORKLOAD_NAME
      value: cmm
    - name: ISTIO_META_OWNER
      value: kubernetes://apis/apps/v1/namespaces/default/deployments/cmm
    - name: ISTIO_META_MESH_ID
      value: cluster.local
    - name: TRUST_DOMAIN
      value: cluster.local
    image: docker.io/istio/proxyv2:1.12.2
    imagePullPolicy: IfNotPresent
    name: istio-proxy
    ports:
    - containerPort: 15090
      name: http-envoy-prom
      protocol: TCP
    readinessProbe:
      failureThreshold: 30
      httpGet:
        path: /healthz/ready
        port: 15021
        scheme: HTTP
      initialDelaySeconds: 1
      periodSeconds: 2
      successThreshold: 1
      timeoutSeconds: 3
    resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: 10m
        memory: 40Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsGroup: 1337
      runAsNonRoot: true
      runAsUser: 1337
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/istio
      name: istiod-ca-cert
    - mountPath: /var/lib/istio/data
      name: istio-data
    - mountPath: /etc/istio/proxy
      name: istio-envoy
    - mountPath: /var/run/secrets/tokens
      name: istio-token
    - mountPath: /etc/istio/pod
      name: istio-podinfo
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-cs7l6
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  initContainers:
  - args:
    - istio-iptables
    - -p
    - "15001"
    - -z
    - "15006"
    - -u
    - "1337"
    - -m
    - REDIRECT
    - -i
    - '*'
    - -x
    - ""
    - -b
    - '*'
    - -d
    - 15090,15021,15020
    - -o
    - "5432"
    image: docker.io/istio/proxyv2:1.12.2
    imagePullPolicy: IfNotPresent
    name: istio-init
    resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: 10m
        memory: 40Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        add:
        - NET_ADMIN
        - NET_RAW
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: false
      runAsGroup: 0
      runAsNonRoot: false
      runAsUser: 0
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-cs7l6
      readOnly: true
  nodeName: minikube
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - emptyDir:
      medium: Memory
    name: istio-envoy
  - emptyDir: {}
    name: istio-data
  - downwardAPI:
      defaultMode: 420
      items:
      - fieldRef:
          apiVersion: v1
          fieldPath: metadata.labels
        path: labels
      - fieldRef:
          apiVersion: v1
          fieldPath: metadata.annotations
        path: annotations
    name: istio-podinfo
  - name: istio-token
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          audience: istio-ca
          expirationSeconds: 43200
          path: istio-token
  - configMap:
      defaultMode: 420
      name: istio-ca-root-cert
    name: istiod-ca-cert
  - name: kube-api-access-cs7l6
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2022-04-16T23:12:42Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2022-04-16T23:12:48Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2022-04-16T23:12:48Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2022-04-16T23:12:40Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://ebf4c8b512c0377fe59e0852b035a625861c331c2c739eaafb7717795e9a67e4
    image: legacymigrate:for_k8
    imageID: docker://sha256:55572d59b0a2eb3d765fb0de090f681012ddf5ca5cc0eaedd2e98ba6ee17e5bc
    lastState:
      terminated:
        containerID: docker://5115833d3e2734fd2e553122ed81c284b7046071daf882f2793aa4c42c70c69b
        exitCode: 1
        finishedAt: "2022-04-16T23:12:46Z"
        reason: Error
        startedAt: "2022-04-16T23:12:42Z"
    name: cmm
    ready: true
    restartCount: 1
    started: true
    state:
      running:
        startedAt: "2022-04-16T23:12:47Z"
  - containerID: docker://480ef50a3506b1f464b31f7886f50cecade849b88916263bb81ac64585a4774d
    image: istio/proxyv2:1.12.2
    imageID: docker-pullable://istio/proxyv2@sha256:f26717efc7f6e0fe928760dd353ed004ea35444f5aa6d41341a003e7610cd26f
    lastState: {}
    name: istio-proxy
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2022-04-16T23:12:42Z"
  hostIP: 192.168.49.2
  initContainerStatuses:
  - containerID: docker://5c5f90fa0a0b2ffdc356f958828d3d33db75b21a006ae97291171ca7a1f841c6
    image: istio/proxyv2:1.12.2
    imageID: docker-pullable://istio/proxyv2@sha256:f26717efc7f6e0fe928760dd353ed004ea35444f5aa6d41341a003e7610cd26f
    lastState: {}
    name: istio-init
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: docker://5c5f90fa0a0b2ffdc356f958828d3d33db75b21a006ae97291171ca7a1f841c6
        exitCode: 0
        finishedAt: "2022-04-16T23:12:41Z"
        reason: Completed
        startedAt: "2022-04-16T23:12:41Z"
  phase: Running
  podIP: 172.17.0.10
  podIPs:
  - ip: 172.17.0.10
  qosClass: Burstable
  startTime: "2022-04-16T23:12:40Z"
```

3)Looking at the iptables rules in the name space of `istio-proxy`, they also look fine from the point of view of the postgresql port(5432):
-------------

The line:
```
 4   240 RETURN     tcp  --  any    any     anywhere             anywhere             tcp dpt:postgresql
```

from full output:
```
root@minikube:~# iptables -t nat -L -v
Chain PREROUTING (policy ACCEPT 682 packets, 40920 bytes)
 pkts bytes target     prot opt in     out     source               destination
  682 40920 ISTIO_INBOUND  tcp  --  any    any     anywhere             anywhere

Chain INPUT (policy ACCEPT 682 packets, 40920 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain OUTPUT (policy ACCEPT 94 packets, 7862 bytes)
 pkts bytes target     prot opt in     out     source               destination
   24  1440 ISTIO_OUTPUT  tcp  --  any    any     anywhere             anywhere

Chain POSTROUTING (policy ACCEPT 101 packets, 8282 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain ISTIO_INBOUND (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 RETURN     tcp  --  any    any     anywhere             anywhere             tcp dpt:15008
    0     0 RETURN     tcp  --  any    any     anywhere             anywhere             tcp dpt:ssh
    0     0 RETURN     tcp  --  any    any     anywhere             anywhere             tcp dpt:15090
  682 40920 RETURN     tcp  --  any    any     anywhere             anywhere             tcp dpt:15021
    0     0 RETURN     tcp  --  any    any     anywhere             anywhere             tcp dpt:15020
    0     0 ISTIO_IN_REDIRECT  tcp  --  any    any     anywhere             anywhere

Chain ISTIO_IN_REDIRECT (3 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 REDIRECT   tcp  --  any    any     anywhere             anywhere             redir ports 15006

Chain ISTIO_OUTPUT (1 references)
 pkts bytes target     prot opt in     out     source               destination
    4   240 RETURN     tcp  --  any    any     anywhere             anywhere             tcp dpt:postgresql
    0     0 RETURN     all  --  any    lo      127.0.0.6            anywhere
    0     0 ISTIO_IN_REDIRECT  all  --  any    lo      anywhere            !localhost
owner UID match 1337
    0     0 RETURN     all  --  any    lo      anywhere             anywhere             ! owner UID match 1337
   13   780 RETURN     all  --  any    any     anywhere             anywhere             owner UID match 1337
    0     0 ISTIO_IN_REDIRECT  all  --  any    lo      anywhere            !localhost
owner GID match 1337
    0     0 RETURN     all  --  any    lo      anywhere             anywhere             ! owner GID match 1337
    0     0 RETURN     all  --  any    any     anywhere             anywhere             owner GID match 1337
    0     0 RETURN     all  --  any    any     anywhere             localhost
    7   420 ISTIO_REDIRECT  all  --  any    any     anywhere             anywhere

Chain ISTIO_REDIRECT (1 references)
 pkts bytes target     prot opt in     out     source               destination
    7   420 REDIRECT   tcp  --  any    any     anywhere             anywhere             redir ports 15001
```

Happy to provide any other details.

### Version

```prose
istioctl version
client version: 1.12.2
control plane version: 1.12.2
data plane version: 1.12.2 (4 proxies)

kubectl version --short
Client Version: v1.19.4
Server Version: v1.22.3
```

**Comment by zirain** [comment]
can you share accesslog from istio-proxy container?

**Comment by livius-ungureanu** [comment]
@zirain  thank you for your reply.

Actually I can not see any logs in the accesslog of the istio-proxy container.
To better isolate the problem, here are the steps to repro the issue with a dummy  http server running on port 5060 in the host machine and a telnet to the dummy server issued from within the a pod container(called sleepubuntu)  that runs in a simple deployment on minikube (minikube installed in the host machine)


1.  `minikube start`
2.  Install istio in the minikube: `istioctl install --set profile=demo -y`
3.  Enable istio auto-injection in the default namespace: `kubectl label namespace default istio-injection=enabled`
4. Add the sleep_ubuntu(here telnet/curl can be installed) in the deployment, where 5060 port is configured to not be proxied by istio-proxy: `kubectl apply -f ./sleep_ubuntu.yaml`
5. Check that proxy side car was created for sleepubuntu. 
```
kubectl get pods
NAME                           READY   STATUS    RESTARTS   AGE
sleepubuntu-66d69dcf7c-6t6kc   2/2     Running   0          111s
```
6. Start the dummy http server on the host machine:
`python3 -m http.server 5060`
7.  Enable accessLogging: `kubectl apply -f telemetri_api.yaml`
8. Step into the sleepubuntu container and issue a telnet to the dummy http server on the host machine:
```
root@sleepubuntu-66d69dcf7c-6t6kc:/# telnet host.minikube.internal 5060
Trying 192.168.49.1...
Connected to host.minikube.internal.
Escape character is '^]'.
```
9. Step into the `istio-proxy` of sleepubuntu and notice the established connection to the dummy http server:
```
kubectl exec -it sleepubuntu-66d69dcf7c-6t6kc -c istio-proxy -- bash
  istio-proxy@sleepubuntu-66d69dcf7c-6t6kc:/$ netstat -an | grep 5060
 tcp        0      0 172.17.0.6:47932        192.168.49.1:5060       ESTABLISHED
```
10. No logs in the access log of the sleep_ubuntu's istio proxy can be observed when telneting/curling the dummy http-server 
```
kubectl logs -l app=sleep_ubuntu -c istio-proxy --tail 1 -f
```
11. Again the iptables rules in the namespace of the `istio-proxy` of `sleepubuntu` looks fine:
```
... 

# dpt:sip is represents 5060 in fact
Chain ISTIO_OUTPUT (1 references)
 pkts bytes target     prot opt in     out     source               destination
   18  1080 RETURN     tcp  --  any    any     anywhere             anywhere             tcp dpt:sip
```







Notes here are the two yaml files involved:
`cat  telemetri_api.yaml`
```
apiVersion: telemetry.istio.io/v1alpha1
kind: Telemetry
metadata:
  name: mesh-default
  namespace: istio-system
spec:
  accessLogging:
    - providers:
      - name: envoy
```

`cat sleep_ubuntu.yaml`
```
##################################################################################################
# Sleep service
##################################################################################################
apiVersion: v1
kind: ServiceAccount
metadata:
  name: sleepubuntu
---
apiVersion: v1
kind: Service
metadata:
  name: sleepubuntu
  labels:
    app: sleepubuntu
    service: sleepubuntu
spec:
  ports:
  - port: 80
    name: http
  selector:
    app: sleepubuntu
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: sleepubuntu
spec:
  replicas: 1
  selector:
    matchLabels:
      app: sleepubuntu
  template:
    metadata:
      labels:
        app: sleepubuntu
      annotations:
          traffic.sidecar.istio.io/excludeOutboundPorts: "5060"
    spec:
      terminationGracePeriodSeconds: 0
      serviceAccountName: sleepubuntu
      containers:
      - name: sleepubuntu
        image: ubuntu
        ports:
        - containerPort: 5060
          name: http5060
        command: ["/bin/sleep", "3650d"]
        imagePullPolicy: IfNotPresent
        volumeMounts:
        - mountPath: /etc/sleep/tls
          name: secret-volume
      volumes:
      - name: secret-volume
        secret:
          secretName: sleep-secret
          optional: true
---
```

**Comment by zirain** [comment]
if `istio-proxy` proxy the traffic, there should be some accesslog records.

**Comment by livius-ungureanu** [comment]
Regarding "if istio-proxy proxy the traffic, there should be some accesslog records", 
I've tried tried the test below.

- On the host machine I've started two dummy servers running on 5060 and 5061 ports. 
   Port 5060 is subject to `traffic.sidecar.istio.io/excludeOutboundPorts`  whereas 5061 not.
-  I've started the accesslog for the istio-proxy of   `sleepubuntu-66d69dcf7c-6t6kc`:  `kubectl logs -l app=sleepubuntu -c istio-proxy --tail 1 -f`
- I've issued two curls from inside the `sleepubuntu-66d69dcf7c-6t6kc` container towards the two dummy servers
  --  curl -sS -v host.minikube.internal:5061
  -- curl -sS -v host.minikube.internal:5060
- In access logs I can see evidence of 5061 traffic but not for 5060 traffic. _This is both are as expected._ so it should indicate that traffic for 5060  is bypassed by istio-proxy.
- But if I telnet those ports at the same time
   ```
  root@sleepubuntu-66d69dcf7c-6t6kc:/# telnet host.minikube.internal 5060
   Trying 192.168.49.1...
  Connected to host.minikube.internal.
   Escape character is '^]'.

  root@sleepubuntu-66d69dcf7c-6t6kc:/# telnet host.minikube.internal 5061
  Trying 192.168.49.1...
  Connected to host.minikube.internal.  
   Escape character is '^]'.
   ```
 Then, I can see both connections from inside  the istio-proxy of `sleepubuntu-66d69dcf7c-6t6kc` when netstated: 
  ```
  tcp        0      0 172.17.0.5:33254        192.168.49.1:5061       ESTABLISHED      // OK
  tcp        0      0 172.17.0.5:53696        192.168.49.1:5060       ESTABLISHED      //  NOK ?!
  ```
 Whereas the connection of 5061  is expected to be seen  from  istio-proxy, I would not expect to see the one of 5060 (bypassed).  Or maybe they the connection of 5060 is shown by netstat inside istio-proxy just because `sleepubuntu-66d69dcf7c-6t6kc` and its  istio-proxy are in the same namespace ?
